pkg/ceph/osd/health.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2017 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package osd
+
+import (
+	"time"
+
+	"github.com/rook/rook/pkg/ceph/client"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/util/proc"
+)
+
+const upStatus = 1
+
+var (
+	healthCheckInterval = 60 * time.Second
+	osdGracePeriod      = 600 * time.Second
+)
+
+// Monitor defines OSD process monitoring
+type Monitor struct {
+	context *clusterd.Context
+	agent   *OsdAgent
+
+	// lastStatus keeps track of OSDs status
+	// key - OSD id; value: time of the status change.
+	lastStatus map[int]time.Time
+}
+
+// NewMonitor instantiates OSD monitoring
+func NewMonitor(context *clusterd.Context, agent *OsdAgent) *Monitor {
+	return &Monitor{context, agent, make(map[int]time.Time)}
+}
+
+// Run runs monitoring logic for osds status at set intervals
+func (m *Monitor) Run() {
+	for {
+		<-time.After(healthCheckInterval)
+		logger.Debug("Checking osd processes status.")
+		err := m.osdStatus()
+		if err != nil {
+			logger.Infof("Failed OSD status check: %s", err)
+		}
+	}
+}
+
+// OSDStatus validates osd dump output
+func (m *Monitor) osdStatus() error {
+	logger.Debugf("OSDs with previously detected Down status: %+v", m.lastStatus)
+	osdDump, err := client.GetOSDDump(m.context, m.agent.cluster.Name)
+	if err != nil {
+		return err
+	}
+
+	evalDownStatus := func(id int, proc *proc.MonitoredProc) {
+		if now := time.Now(); now.Sub(m.lastStatus[id]) > osdGracePeriod {
+			err = proc.Stop()
```

Comment:
probably right before calling this stop, we will want to log that we are going to do the stop.  something like:
```
logger.Infof("stopping osd.%d because it has been down for longer than the grace period (down since %+v)", id, m.lastStatus[id])
```